src/Nethermind/Nethermind.Evm/VirtualMachine.cs

```diff
@@ -1246,12 +1246,73 @@ static void UpdateCurrentState(EvmState state, int pc, long gas, int stackHead)
                     }
                 case Instruction.EXTCODESIZE:
                     {
+                        Metrics.ExtCodeSizeOpcode++;
                         long gasCost = spec.GetExtCodeCost();
                         if (!UpdateGas(gasCost, ref gasAvailable)) goto OutOfGas;
 
                         Address address = stack.PopAddress();
                         if (!ChargeAccountAccessGas(ref gasAvailable, vmState, address, spec)) goto OutOfGas;
 
+                        if (programCounter < code.Length)
+                        {
+                            bool optimizeAccess = false;
+                            Instruction nextInstruction = (Instruction)code[programCounter];
+                            if (nextInstruction == Instruction.ISZERO)
+                            {
+                                optimizeAccess = true;
+                                Metrics.ExtCodeSizeOptimizedIsZero++;
+                            }
+                            else if ((nextInstruction == Instruction.GT || nextInstruction == Instruction.EQ) &&
+                                    stack.PeekUInt256IsZero())
+                            {
+                                optimizeAccess = true;
+                                stack.PopLimbo();
+                                if (nextInstruction == Instruction.GT)
+                                {
+                                    Metrics.ExtCodeSizeOptimizedGT++;
+                                }
+                                else if (nextInstruction == Instruction.EQ)
+                                {
+                                    Metrics.ExtCodeSizeOptimizedEQ++;
+                                }
+                            }
+
+                            if (optimizeAccess)
+                            {
+                                // EXTCODESIZE ISZERO/GT/EQ peephole optimization.
+                                // In solidity 0.8.1+: `return account.code.length > 0;`
+                                // is is a common pattern to check if address is a contract
+                                // however we can just check the address's loaded CodeHash
+                                // to reduce storage access from trying to load the code
+                                if (traceOpcodes)
+                                {
+                                    EndInstructionTrace(gasAvailable, vmState.Memory?.Size ?? 0);
+                                    StartInstructionTrace(Instruction.ISZERO, vmState, gasAvailable, programCounter, in stack);
```

Comment:
This trace us wrong I believe as you always trace `Instruction.ISZERO`.

Reply:
Is same as all, only traces at start of `while` based on `traceOpcodes`; isn't any special trace in `case ISZERO:`?

Reply:
The trace is wrong -- since you omit adding the acual value to the stack. 
```
{"pc":25,"op":59,"gas":"0x79ba1c","gasCost":"0xa28","memSize":0,"stack":["0x24","0x44","0x74","0xd1","0x6b","0xb9","0x33","0x866833515b6d086c607f"],"depth":1,"refund":0,"opname":"EXTCODESIZE","error":""}
{"pc":26,"op":21,"gas":"0x79aff4","gasCost":"0x3","memSize":0,"stack":["0x24","0x44","0x74","0xd1","0x6b","0xb9","0x33"],"depth":1,"refund":0,"opname":"ISZERO","error":""}
{"pc":27,"op":116,"gas":"0x79aff1","gasCost":"0x3","memSize":0,"stack":["0x24","0x44","0x74","0xd1","0x6b","0xb9","0x33","0x1"],"depth":1,"refund":0,"opname":"PUSH21","error":""}
```
The `EXTCODESIZE` consumes `0x866833515b6d086c607f`, leaving `0x33` at the top of the stack, making it look lik e `ISZERO` operates on `0x33`. While in fact it operates on a value that never is put onto the stack